languageServer/src/DvcTextDocument.ts

```diff
@@ -0,0 +1,319 @@
+    if (!this.uri.endsWith('dvc.yaml')) {
+      return null
+    }
+
+    const codeActions: CodeAction[] = []
+
+    const model = new DvcYaml(this)
+    const stage = model.getStageAt(params.range.start)
+    if (stage?.cmd) {
+      const filesUsed = stage.cmd.getReferencedFiles()
+      const missingDeps = filesUsed.filter(path => !stage.deps?.has(path))
+
+      if (missingDeps.length > 0) {
+        stage.addDependencies(missingDeps)
+        const yamlSource = stage.parentDvcYaml.toString()
+        const textEdits = this.calculateTextEdits(yamlSource)
+
+        const workspaceEdit: WorkspaceEdit = {
+          changes: {
+            [this.uri]: textEdits
+          }
+        }
+
+        const codeAction = CodeAction.create(
+          'Add cmd files as dependencies',
+          workspaceEdit,
+          CodeActionKind.RefactorRewrite
+        )
+
+        codeActions.push(codeAction)
+      }
+    }
+
+    return codeActions
+  }
+
+  convertYamlRange(yamlRange: [number, number, number]) {
+    const symbolStart = yamlRange[0]
+    const valueEnd = yamlRange[1]
+    const symbolEnd = yamlRange[2]
+    const range = Range.create(
+      this.positionAt(symbolStart),
+      this.positionAt(symbolEnd)
+    )
+    const selectionRange = Range.create(
+      this.positionAt(symbolStart),
+      this.positionAt(valueEnd)
+    )
+
+    return {
+      range,
+      selectionRange
+    }
+  }
+
+  yamlNodeToDocumentSymbols(
+    node: Node | Pair,
+    range: [number, number, number]
+  ): DocumentSymbol[] {
+    if (isNode(node)) {
+      if (isScalar(node)) {
+        return this.yamlScalarNodeToDocumentSymbols(node, range)
+      }
+    } else if (isPair(node)) {
+      return this.yamlNodeToDocumentSymbols(node.value as Node | Pair, range)
+    }
+
+    return []
+  }
+
+  symbolScopeAt(position: Position): DocumentSymbol[] {
+    const cursorOffset: number = this.offsetAt(position)
+
+    const symbolsFound: Array<DocumentSymbol | null> = []
+
+    visit(this.getYamlDocument(), (_key, node) => {
+      if (isNode(node)) {
+        const range = node.range
+        if (!!range && cursorOffset >= range[0] && cursorOffset <= range[2]) {
+          symbolsFound.push(...this.yamlNodeToDocumentSymbols(node, range))
+        }
+      }
+    })
+    const symbolStack = (symbolsFound.filter(Boolean) as DocumentSymbol[]).sort(
+      (a, b) => {
+        const offA = this.offsetAt(a.range.end) - this.offsetAt(a.range.start)
+        const offB = this.offsetAt(b.range.end) - this.offsetAt(b.range.start)
+
+        return offB - offA // We want the tighter fits for last, so we can just pop them
+      }
+    )
+
+    return [...symbolStack]
+  }
+
+  symbolAt(position: Position): DocumentSymbol | undefined {
+    return this.symbolScopeAt(position).pop()
+  }
+
+  getTextDocuments() {
+    return this.documents
+  }
+
+  createFinder(txtDoc: TextDocument) {
+    return new DvcTextDocument(txtDoc, this.documents)
```

Comment:
geTextDocuments is right there, any other usage you could replace? Or just drop the accessor?